torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+//
+// The implementation is based off of the following paper:
+// "A Dynamic Topsort Algorithm for Directed Acyclic Graphs"
+// by David Pearce and Paul Kelly
+// https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf
+
+template <typename T> struct Vertex;
+
+template <typename T>
+using vertex_set = std::unordered_set<Vertex<T>*>;
+
+template <typename T>
+using vertex_list = std::vector<Vertex<T>*>;
+
+template <typename T>
+using maybe_vertex_list = std::vector<at::optional<Vertex<T>*>>;
+
+template <typename T>
+struct Vertex {
+  std::vector<T> rdata; // stored in reverse order
+  size_t ord; // index in topological ordering. Is unique.
+  bool visited; // If this vertex has been visited
+  vertex_set<T> in_edges;
+  vertex_set<T> out_edges;
+
+  std::string toString();
+};
+
+template <typename T>
+struct DynamicDAG {
+  ~DynamicDAG();
+
+  Vertex<T>* newVertex(T datum);
+  std::pair<vertex_set<T>,vertex_set<T>> removeVertex(Vertex<T>* v);
+
+  void addEdge(Vertex<T>* producer, Vertex<T>* consumer);
+  void removeEdge(Vertex<T>* producer, Vertex<T>* consumer);
+  bool contractEdge(Vertex<T>* producer, Vertex<T>* consumer);
+
+  bool contractionProducesCycle(Vertex<T>* producer, Vertex<T>* consumer);
+  bool dfsForward(Vertex<T>* producer, size_t upper_bound, vertex_list<T>& visited);
+  void dfsBackward(Vertex<T>* producer, size_t lower_bound, vertex_list<T>& visited);
+  void clearVisited(vertex_list<T>& visited);
+
+  void reorder(vertex_list<T>& deltaF, vertex_list<T>& deltaB);
+  void sort(vertex_list<T>& delta);
+
+  void checkInvariants();
+
+  Vertex<T>* at(size_t ord) const { return vertices_.at(ord).value(); };
+  const maybe_vertex_list<T>& maybe_vertices() const { return vertices_; };
+
+  // Use for debugging. Don't call this often.
+  size_t numVertices() const;
+
+  std::string toString();
+
+ private:
+
+  // Store vertices indexed by their topological order.
+  // If a vertex v has ord 5, then it can be found at vertices_[5].
+  // There may be gaps in vertices_; this is to enable fast deletion.
+  maybe_vertex_list<T> vertices_;
+
+  // temp buffers used in the PK algorithm.
+  // Their usage makes this data structure NOT thread safe.
+  vertex_list<T> deltaF_;  // all vertices visited by dfsForward
+  vertex_list<T> deltaB_;  // all vertices visited by dfsBackward
+};
+
+template <typename T>
+DynamicDAG<T>::~DynamicDAG() {
+  for (auto v : vertices_) {
+    if (v.has_value()) {
+      delete v.value();
+    }
+  }
+}
+
+// O(vertices_.size()). Used for testing, don't call this often.
+template <typename T>
+size_t DynamicDAG<T>::numVertices() const {
+  return std::count_if(vertices_.begin(), vertices_.end(), [](at::optional<Vertex<T>*> v) {
+      return v;
+  });
+}
+
+template <typename T>
+Vertex<T>* DynamicDAG<T>::newVertex(T datum) {
+  auto* vertex = new Vertex<T>();
+  vertex->rdata.push_back(datum);
+  vertex->visited = false;
+  vertex->ord = vertices_.size();
```

Comment:
Can we make a constructor for `Vertex` instead? That would make it harder to forget to initialize a member.